workflow/scripts/add_district_heating_subnodes.py

```diff
@@ -0,0 +1,421 @@
+import pypsa
+import xarray as xr
+from typing import Union
+
+
+def prepare_subnodes(
+    subnodes: pd.DataFrame,
+    cities: gpd.GeoDataFrame,
+    regions_onshore: gpd.GeoDataFrame,
+    lau: gpd.GeoDataFrame,
+    head: Union[int, bool] = 40,
+) -> gpd.GeoDataFrame:
+    """
+    Prepare subnodes by filtering district heating systems data for largest systems and assigning the corresponding LAU and onshore region shapes.
+
+    Parameters:
+    subnodes (pd.DataFrame): DataFrame containing information about district heating systems.
+    cities (gpd.GeoDataFrame): GeoDataFrame containing city coordinates with columns 'Stadt' and 'geometry'.
+    regions_onshore (gpd.GeoDataFrame): GeoDataFrame containing onshore region geometries of clustered network.
+    lau (gpd.GeoDataFrame): GeoDataFrame containing LAU (Local Administrative Units) geometries and IDs.
+    heat_techs (gpd.GeoDataFrame): GeoDataFrame containing heat technology geometries.
+    head (Union[int, bool], optional): Number of largest district heating networks to keep. Defaults to 40. If set to True, it will be set to 40.
+
+    Returns:
+    gpd.GeoDataFrame: GeoDataFrame with processed subnodes, including geometries, clusters, LAU IDs, and NUTS3 shapes.
+    """
+    # If head is boolean set it to 40 for default behavior
+    if isinstance(head, bool):
+        head = 40
+
+    subnodes["Stadt"] = subnodes["Stadt"].str.split("_").str[0]
+
+    # Drop duplicates if Gelsenkirchen, Kiel, or Flensburg is included and keep the one with higher Wärmeeinspeisung in GWh/a
+    subnodes = subnodes.drop_duplicates(subset="Stadt", keep="first")
+
+    # Keep only n largest district heating networks according to head parameter
+    subnodes = subnodes.sort_values(
+        by="Wärmeeinspeisung in GWh/a", ascending=False
+    ).head(head)
+
+    subnodes["yearly_heat_demand_MWh"] = subnodes["Wärmeeinspeisung in GWh/a"] * 1e3
+
+    logger.info(
+        f"The selected district heating networks have an overall yearly heat demand of {subnodes['yearly_heat_demand_MWh'].sum()} MWh/a. "
+    )
+
+    subnodes["geometry"] = subnodes["Stadt"].apply(
+        lambda s: cities.loc[cities["Stadt"] == s, "geometry"].values[0]
+    )
+
+    subnodes = subnodes.dropna(subset=["geometry"])
+
+    # Convert the DataFrame to a GeoDataFrame
+    subnodes = gpd.GeoDataFrame(subnodes, crs="EPSG:4326")
+
+    # Assign cluster to subnodes according to onshore regions
+    subnodes["cluster"] = subnodes.apply(
+        lambda x: regions_onshore.geometry.contains(x.geometry).idxmax(), axis=1
+    )
+    subnodes["lau"] = subnodes.apply(
+        lambda x: lau.loc[lau.geometry.contains(x.geometry).idxmax(), "LAU_ID"], axis=1
+    )
+    subnodes["lau_shape"] = subnodes.apply(
+        lambda x: lau.loc[lau.geometry.contains(x.geometry).idxmax(), "geometry"].wkt,
+        axis=1,
+    )
+
+    return subnodes
+
+
+def add_subnodes(n: pypsa.Network, subnodes: gpd.GeoDataFrame) -> None:
+    """
```

Comment:
Could we change the docstring style to numpy?

Reply:
Done